controllers/metal3.io/host_state_machine.go

```diff
@@ -203,39 +220,17 @@ func (hsm *hostStateMachine) handleUnmanaged(info *reconcileInfo) actionResult {
 }
 
 func (hsm *hostStateMachine) handleRegistering(info *reconcileInfo) actionResult {
-	actResult := hsm.Reconciler.actionRegistering(hsm.Provisioner, info)
-
-	switch actResult.(type) {
-	case actionComplete:
-		// TODO: In future this state should only occur before the host is
-		// registered the first time (though we must always check and
-		// re-register the host regardless of the current state). That will
-		// eliminate the need to determine which state we came from here.
-		switch {
-		case hsm.Host.Spec.ExternallyProvisioned:
-			hsm.NextState = metal3v1alpha1.StateExternallyProvisioned
-		case hsm.Host.WasProvisioned():
-			hsm.NextState = metal3v1alpha1.StateProvisioned
-		case hsm.Host.NeedsHardwareInspection():
-			hsm.NextState = metal3v1alpha1.StateInspecting
-		case hsm.Host.NeedsHardwareProfile():
-			hsm.NextState = metal3v1alpha1.StateMatchProfile
-		default:
-			hsm.NextState = metal3v1alpha1.StateReady
-		}
-	case actionFailed:
-		hsm.NextState = metal3v1alpha1.StateRegistrationError
-	}
-	return actResult
-}
-
-func (hsm *hostStateMachine) handleRegistrationError(info *reconcileInfo) actionResult {
-	if !hsm.Host.Status.TriedCredentials.Match(*info.bmcCredsSecret) {
-		info.log.Info("Modified credentials detected; will retry registration")
-		hsm.NextState = metal3v1alpha1.StateRegistering
-		return actionComplete{}
+	// Getting to the state handler at all means we have successfully
+	// registered using the current BMC credentials, so we can move to the
+	// next state. We will not return to the Registering state, even
+	// if the credentials change and the Host must be re-registered.
+	hsm.Host.ClearError()
```

Comment:
See earlier comment about clearing errors.

Reply:
This is a bit of a special case because we don't call an action*() function from this state any more. I'm not sure it's even necessary, but I wrote this patch a year ago so I might have known the reason then :D